File: samples/cpp/inpaint.cpp
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/photo.hpp"

#include <iostream>

using namespace cv;
using namespace std;

static void help(char** argv)
{
    cout << "\nCool inpainting demo. Inpainting repairs damage to images by floodfilling the damage \n"
            << "with surrounding image areas.\n"
            "Using OpenCV version " << CV_VERSION << "\n"
            "Usage:\n" << argv[0] << " [image_name -- Default fruits.jpg]\n" << endl;
}

int main(int argc, char** argv)
{
    cv::CommandLineParser parser(argc, argv, "{@image|fruits.jpg|}");
    help(argv);

    string filename = samples::findFile(parser.get<string>("@image"));
    Mat img0 = imread(filename, IMREAD_COLOR);
    if (img0.empty())
    {
        cout << "Couldn't open the image " << filename << ". Usage: inpaint <image_name>\n" << endl;
        return 0;
    }

    Mat img = img0.clone();
    Mat inpaintMask = Mat::zeros(img.size(), CV_8U);

    // Simulate drawing on the mask for demo purposes
    rectangle(inpaintMask, Point(50, 50), Point(150, 150), Scalar::all(255), FILLED);
    rectangle(img, Point(50, 50), Point(150, 150), Scalar::all(255), FILLED);

    // Run inpainting algorithm
    Mat inpainted;
    inpaint(img, inpaintMask, inpainted, 3, INPAINT_TELEA);

    // Save the inpainted image
    imwrite("inpainted_image.jpg", inpainted);
    cout << "Inpainted image saved as: inpainted_image.jpg" << endl;

    return 0;
}

File: samples/python/inpaint.py
#!/usr/bin/env python

'''
Inpainting sample.

Inpainting repairs damage to images by floodfilling
the damage with surrounding image areas.

Usage:
  inpaint.py [<image>]

Keys:
  SPACE - inpaint
  r     - reset the inpainting mask
  ESC   - exit
'''

# Python 2/3 compatibility
from __future__ import print_function

import numpy as np
import cv2 as cv

from common import Sketcher

def main():
    import sys
    try:
        fn = sys.argv[1]
    except:
        fn = 'fruits.jpg'

    img = cv.imread(cv.samples.findFile(fn))
    if img is None:
        print('Failed to load image file:', fn)
        sys.exit(1)

    img_mark = img.copy()
    mark = np.zeros(img.shape[:2], np.uint8)
    sketch = Sketcher('img', [img_mark, mark], lambda : ((255, 255, 255), 255))

    while True:
        ch = cv.waitKey()
        if ch == 27:
            break
        if ch == ord(' '):
            res = cv.inpaint(img_mark, mark, 3, cv.INPAINT_TELEA)
            cv.imshow('inpaint', res)
        if ch == ord('r'):
            img_mark[:] = img
            mark[:] = 0
            sketch.show()

    print('Done')


if __name__ == '__main__':
    print(__doc__)
    main()
    cv.destroyAllWindows()
